### ISSUE TITLE
[Bug][Validator] Validating against a collection does not work with latest doctrine/collections

### ISSUE THREAD

**Issue by jmsche** [symfony/symfony#30685]
**Symfony version(s) affected**: 4.2.4

**Description**  
Hi, I upgraded some project this morning, and when I try to validate an entity with a collection in it, it fails.
Seems to be due to latest doctrine/collections release (1.6.0), worked well with previous ones.

**How to reproduce**  
Submit a form and validate it; the validated object must have a collection in it which also must be validated (using the `@Assert\Valid()` annotation.

Here is a reproducer: https://github.com/jmsche/annotation-reproducer
Run `composer install` open the default page and click the "Submit button".

Here is the error:

> [Semantical Error] The annotation "@template" in class Doctrine\Common\Collections\ArrayCollection was never imported. Did you maybe forget to add a "use" statement for this annotation?

And the stack trace:

```
Doctrine\Common\Annotations\AnnotationException:
[Semantical Error] The annotation "@template" in class Doctrine\Common\Collections\ArrayCollection was never imported. Did you maybe forget to add a "use" statement for this annotation?

  at vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/AnnotationException.php:54
  at Doctrine\Common\Annotations\AnnotationException::semanticalError('The annotation &quot;@template&quot; in class Doctrine\\Common\\Collections\\ArrayCollection was never imported. Did you maybe forget to add a &quot;use&quot; statement for this annotation?')
     (vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/DocParser.php:727)
  at Doctrine\Common\Annotations\DocParser->Annotation()
     (vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/DocParser.php:663)
  at Doctrine\Common\Annotations\DocParser->Annotations()
     (vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/DocParser.php:354)
  at Doctrine\Common\Annotations\DocParser->parse('/** * An ArrayCollection is a Collection implementation that wraps a regular PHP array. * * Warning: Using (un-)serialize() on a collection is not a supported use-case * and may break when we change the internals in the future. If you need to * serialize a collection use {@link toArray()} and reconstruct the collection * manually. * * @template TKey of array-key * @template T * @template-implements Collection&lt;TKey,T&gt; * @template-implements Selectable&lt;TKey,T&gt; */', 'class Doctrine\\Common\\Collections\\ArrayCollection')
     (vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/AnnotationReader.php:222)
  at Doctrine\Common\Annotations\AnnotationReader->getClassAnnotations(object(ReflectionClass))
     (vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/CachedReader.php:79)
  at Doctrine\Common\Annotations\CachedReader->getClassAnnotations(object(ReflectionClass))
     (vendor/symfony/validator/Mapping/Loader/AnnotationLoader.php:45)
  at Symfony\Component\Validator\Mapping\Loader\AnnotationLoader->loadClassMetadata(object(ClassMetadata))
     (vendor/symfony/validator/Mapping/Loader/LoaderChain.php:54)
  at Symfony\Component\Validator\Mapping\Loader\LoaderChain->loadClassMetadata(object(ClassMetadata))
     (vendor/symfony/validator/Mapping/Factory/LazyLoadingMetadataFactory.php:105)
  at Symfony\Component\Validator\Mapping\Factory\LazyLoadingMetadataFactory->getMetadataFor(object(ArrayCollection))
     (vendor/symfony/validator/Validator/RecursiveContextualValidator.php:316)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateObject(object(ArrayCollection), 'data.tags', array('Default'), 1, object(ExecutionContext))
     (vendor/symfony/validator/Validator/RecursiveContextualValidator.php:698)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateGenericNode(object(ArrayCollection), object(Article), '00000000099a8bfe0000000077fc4d49:App\\Entity\\Article:tags', object(PropertyMetadata), 'data.tags', array('Default'), array('Default'), 1, object(ExecutionContext))
     (vendor/symfony/validator/Validator/RecursiveContextualValidator.php:547)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateClassNode(object(Article), '00000000099a8bfe0000000077fc4d49', object(ClassMetadata), 'data', array('Default'), null, 1, object(ExecutionContext))
     (vendor/symfony/validator/Validator/RecursiveContextualValidator.php:323)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateObject(object(Article), 'data', array('Default'), 1, object(ExecutionContext))
     (vendor/symfony/validator/Validator/RecursiveContextualValidator.php:141)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate(object(Article), null, array('Default'))
     (vendor/symfony/form/Extension/Validator/Constraints/FormValidator.php:52)
  at Symfony\Component\Form\Extension\Validator\Constraints\FormValidator->validate(object(Form), object(Form))
     (vendor/symfony/validator/Validator/RecursiveContextualValidator.php:809)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateInGroup(object(Form), '00000000099a8a0f0000000077fc4d49', object(ClassMetadata), 'Default', object(ExecutionContext))
     (vendor/symfony/validator/Validator/RecursiveContextualValidator.php:525)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateClassNode(object(Form), '00000000099a8a0f0000000077fc4d49', object(ClassMetadata), '', array('Default'), null, 1, object(ExecutionContext))
     (vendor/symfony/validator/Validator/RecursiveContextualValidator.php:323)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateObject(object(Form), '', array('Default'), 1, object(ExecutionContext))
     (vendor/symfony/validator/Validator/RecursiveContextualValidator.php:141)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate(object(Form), null, array('Default'))
     (vendor/symfony/validator/Validator/RecursiveValidator.php:100)
  at Symfony\Component\Validator\Validator\RecursiveValidator->validate(object(Form))
     (vendor/symfony/form/Extension/Validator/EventListener/ValidationListener.php:55)
  at Symfony\Component\Form\Extension\Validator\EventListener\ValidationListener->validateForm(object(FormEvent), 'form.post_submit', object(EventDispatcher))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:212)
  at Symfony\Component\EventDispatcher\EventDispatcher->doDispatch(array(array(object(ValidationListener), 'validateForm')), 'form.post_submit', object(FormEvent))
     (vendor/symfony/event-dispatcher/EventDispatcher.php:44)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch('form.post_submit', object(FormEvent))
     (vendor/symfony/event-dispatcher/ImmutableEventDispatcher.php:33)
  at Symfony\Component\EventDispatcher\ImmutableEventDispatcher->dispatch('form.post_submit', object(FormEvent))
     (vendor/symfony/form/Form.php:673)
  at Symfony\Component\Form\Form->submit(array(), true)
     (vendor/symfony/form/Extension/HttpFoundation/HttpFoundationRequestHandler.php:108)
  at Symfony\Component\Form\Extension\HttpFoundation\HttpFoundationRequestHandler->handleRequest(object(Form), object(Request))
     (vendor/symfony/form/Form.php:494)
  at Symfony\Component\Form\Form->handleRequest(object(Request))
     (src/Controller/FormController.php:21)
  at App\Controller\FormController->__invoke(object(Request))
     (vendor/symfony/http-kernel/HttpKernel.php:150)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:67)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:198)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:25)
```

**Possible Solution**  
N/A

**Additional context**  
N/A

**Comment by mmarton** [comment]
Fixed in doctrine/collections@1.6.1
https://github.com/doctrine/collections/pull/186

it was a problem between frameworkExtraBundle's @Template annotation and doctrine's @template typehint for psalm.

**Comment by xabbuh** [comment]
I am closing here as there is indeed nothing that we can do, but this was fixed on the Doctrine side.
